[WIP] DEP remove lars positive parameter

XavierSATTLER opened the pull request:
#### Reference Issues/PRs
#13863

#### What does this implement/fix? Explain your changes.
I removed the positive argument from the lars functions and adapted the tests + documentation

#### Any other comments?
*Issue to deal with before merging this pull request* :
in decomposition.dict_learning, some functions use lars algorithm by default (namely : `sparse_encode`, `_sparse_encode`, `dict_learning`, `dict_learning_online`, `MiniBatchDictionaryLearning.__init__`).

However, they can also use other algorithms as well, and therefore still have a `positive` parameter.

Using the positive parameter with those functions without changing the algorithm will throw an error, (example: `sparse_encode(X, dictionary, positive=True)`) which looks weird to me..

What do you think is the right thing to do ?

1. Changing the default value (hence, people migrating their code won't have any error, but also won't have any clue the same call to the same function will have different behaviour on different versions of sklearn)
2. Letting the code crash for the ones who used positive parameters with default lars algorithm in their code ?

XavierSATTLER committed:
removed lars positive parameter

XavierSATTLER committed:
removed positive argument in tests

XavierSATTLER committed:
refactored tests to be more respectful of testing framework

XavierSATTLER committed:
cleaned one other test

XavierSATTLER committed:
flake8

XavierSATTLER committed:
added lars positive parameter tests

glemaitre commented on sklearn/decomposition/tests/test_dict_learning.py:
```diff
@@ -192,10 +210,18 @@ def test_dict_learning_online_positivity(transform_algorithm,
     n_components = 8
 
-    dico = MiniBatchDictionaryLearning(
+    clf = MiniBatchDictionaryLearning(
```
dico was better than clf. This is not a classifier

glemaitre commented on sklearn/decomposition/tests/test_dict_learning.py:
```diff
@@ -76,10 +82,18 @@ def test_dict_learning_positivity(transform_algorithm,
                                   positive_dict):
     n_components = 5
-    dico = DictionaryLearning(
+    clf = DictionaryLearning(
```
revert the change of variable name

glemaitre commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -729,6 +752,12 @@ def dict_learning_online(X, n_components=2, alpha=1, n_iter=100,
+
+    if method == 'lars' and positive_code:
+        raise ValueError(
+            "Positive constraint not supported for \"lars\" coding method."
```
```suggestion
            "Positive constraint not supported for 'lars' coding method."
```

glemaitre commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -143,11 +151,20 @@ def _sparse_encode(X, dictionary, gram, cov=None, algorithm='lasso_lars',
         try:
             err_mgt = np.seterr(all='ignore')
 
+            # TODO: Should verbose argument be passed to this?
```
You can remove this

glemaitre commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -112,12 +112,20 @@ def _sparse_encode(X, dictionary, gram, cov=None, algorithm='lasso_lars',
+            if positive:
+                raise ValueError(
+                    """
+                    Positive constraint not supported for \"lasso_lars\"
```
Actually, `positive` was not deprecated in `lasso_lars` for what I can see in the deprecation message.
@agramfort could you confirm which class is affected by the `positive` keyword deprecation?

XavierSATTLER replied:
`LassoLars` class inherits from `Lars` class (only), so when I removed `positive` from the `Lars` class, I removed it from `LassoLars` as well.

I looked at the source code, searching for LassoLars reimplementation of the `positive` parameter and couldn't find any.

However, as I am a pretty recent contributor, I may have missed something and my comment may need a double check.

glemaitre replied:
I spoke with @agramfort IRL and we should not remove the `positive` parameter from the `LassoLars` since we force it to use `lasso` as a method (this is hardcoded in the class).

I already implemented the trick in the previous PR which I closed #13782 
The idea is that you can set a variable `_positive` to be `False` for the `Lars` class and not expose the parameter:

https://github.com/scikit-learn/scikit-learn/pull/13782/files#diff-bdf3cc726bb0cd772cd2ffd870e2ece8R851

Then you can check the `_fit` method as well.

XavierSATTLER replied:
I re-added the positive parameter in lasso-lars algorithm.

I actually set the variable `positive` to `False` in `Lars` class (and not `_positive`).

Using a `_positive` variable would have forced me to use `self._positive` in every algorithm implemented in `Lars`. As `LassoLars` inherits from `Lars`, I would have had to do someting like

```
class LassoLars(Lars):
     def __init____(self, [ ... ], positive):
        ...
        self._positive = positive
```

glemaitre commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -519,6 +536,12 @@ def dict_learning(X, n_components, alpha, max_iter=100, tol=1e-8,
+
+    if method == 'lars' and positive_code:
+        raise ValueError(
+            "Positive constraint not supported for \"lars\" coding method."
```
```suggestion
            "Positive constraint not supported for 'lars' coding method."
```

glemaitre commented on sklearn/decomposition/tests/test_dict_learning.py:
```diff
@@ -56,8 +57,13 @@ def test_dict_learning_overcomplete():
+def test_dict_learning_lars_positive_parameter():
+    n_components = 5
+    alpha = 1
+    assert_raises(ValueError, dict_learning, X, n_components,
```
We are using `pytest.raises` form now on. You should also match the error message.

glemaitre commented on sklearn/decomposition/tests/test_dict_learning.py:
```diff
@@ -76,10 +82,18 @@ def test_dict_learning_positivity(transform_algorithm,
+        fit_algorithm="cd")
+
+    dico = clf.fit(X)
+    if transform_algorithm in ["lasso_lars", "lars"] and positive_code:
```
We should change the parametrization instead of changing the tests.

glemaitre commented on sklearn/decomposition/tests/test_dict_learning.py:
```diff
@@ -170,8 +184,12 @@ def test_dict_learning_online_shapes():
-@pytest.mark.filterwarnings('ignore::DeprecationWarning')
+def test_dict_learning_online_lars_positive_parameter():
+    alpha = 1
+    assert_raises(ValueError, dict_learning_online, X,
```
use pytest

glemaitre commented on sklearn/decomposition/tests/test_dict_learning.py:
```diff
@@ -192,10 +210,18 @@ def test_dict_learning_online_positivity(transform_algorithm,
+        fit_algorithm='cd')
+
+    dico = clf.fit(X)
+    if transform_algorithm in ["lasso_lars", "lars"] and positive_code:
```
change the decorator

glemaitre commented on sklearn/decomposition/tests/test_dict_learning.py:
```diff
@@ -318,18 +343,17 @@ def test_sparse_encode_positivity(positive):
-    except ValueError:
-        if not positive:
-            raise
+    for algo in ('lasso_lars', 'lars', 'omp'):
```
I would split the test into 2 tests. One that should raise error for not supporting the parameter and another checking that the coefficient are positive

XavierSATTLER committed:
replaced \" by ' inside a string

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

XavierSATTLER committed:
replaced \" by ' inside a string

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

XavierSATTLER committed:
Removed TODO comment

XavierSATTLER committed:
used pytest.raise in test, matching err_msg

XavierSATTLER committed:
revert dico to clf name change

XavierSATTLER committed:
removed TODO comment

XavierSATTLER committed:
replaced " by ' in string

XavierSATTLER committed:
Updated tests parametrization and used pytest

XavierSATTLER committed:
updated string value errors and split test_sparse_encode_positivity into 2 tests

XavierSATTLER committed:
split test_dict_learning_online_positivity test into 3 coherent tests

XavierSATTLER committed:
removed variable that could be used directly

XavierSATTLER commented:
Hi @glemaitre ,

Thank you for all your comments. I took them all into account, thinking they would be moved instead of removed...

Removing the `positive` parameter was actually a bit tricky from a test perspective.
I am telling you this because I find it makes the diffs pretty hard to read.

Before, it was quite straightforward to use `@pytest.mark.parametrize` on test functions with `positive_code` either set to True or False. However, removing it led to the need of splitting tests into different functions (in order to avoid having case check in the functions). Therefore, I ended up splitting one test function into three. (one for the situations still working independant of the `positive` parameter changes, one for the situations still working for lars method and one for raising errors).

For the rest, I refactored the tests using pytest as you asked and changed the way the error strings were built.

agramfort commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -729,6 +744,12 @@ def dict_learning_online(X, n_components=2, alpha=1, n_iter=100,
+    if method == 'lars' and positive_code:
+        raise ValueError(
+            "Positive constraint not supported for 'lars' coding method."
+        )
```
rather than duplicating this error message can you make a private function. Something like

`_check_positive_coding(method, positive)`

that would crash if needed.

XavierSATTLER committed:
Put positive coding error in a private function

jnothman commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -519,6 +523,12 @@ def dict_learning(X, n_components, alpha, max_iter=100, tol=1e-8,
         raise ValueError('Coding method %r not supported as a fit algorithm.'
                          % method)
+
+    if method == 'lars' and positive_code:
```
Can't you reuse `_check_positive_coding`?

XavierSATTLER committed:
replaced value error check by _check_positive_coding

jnothman commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -178,6 +174,14 @@ def _sparse_encode(X, dictionary, gram, cov=None, algorithm='lasso_lars',
     return new_code
 
 
+def _check_positive_coding(method, positive):
```
move this to the top of the file, after imports but before public API

jnothman commented on sklearn/decomposition/tests/test_dict_learning.py:
```diff
@@ -205,7 +251,44 @@ def test_dict_learning_online_positivity(transform_algorithm,
+    False,
+    True,
+])
+def test_minibatch_dictionnary_learning_lars(transform_algorithm,
```
dictionnary -> dictionary

jnothman reviewed (approved):
Otherwise LGTM

XavierSATTLER committed:
Moved _check_positive_coding on top of the file

XavierSATTLER committed:
Fixed typo in dictionary

XavierSATTLER committed:
re-added positive parameter for lassolars in doc

XavierSATTLER committed:
Removed lasso_lars from _check_positive_coding

XavierSATTLER committed:
re-added positive parameter for lasso lars in _sparse_encode

XavierSATTLER committed:
reset tests to accept positive parameter with lasso_lars algorithm

XavierSATTLER committed:
re-added positive parameter to lasso_lars

XavierSATTLER committed:
Fixed errors while re-adding positive on lasso_lars

XavierSATTLER committed:
fixed doc

XavierSATTLER committed:
moved documentation

XavierSATTLER committed:
removed useless blank line

XavierSATTLER committed:
updated Lars class to have self.positive=False by default

XavierSATTLER committed:
moved positiion where positive is defined in lars

XavierSATTLER committed:
reset test_estimatorclasses_positive_constraint to its original writing

XavierSATTLER committed:
re-added warning filter in tests

glemaitre committed:
nitpicks

glemaitre commented:
Thanks @XavierSATTLER!!!
